Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -17,296 +17,516 @@
+
+    // Image Drawing
+    D2D1_INTERPOLATION_MODE bitmapInterpolationMode = D2D1_INTERPOLATION_MODE_NEAREST_NEIGHBOR;
+
+    // Userspace Coordinate Transformation
+    CGAffineTransform transform{ CGAffineTransformIdentity };
+    CGAffineTransform textMatrix{ CGAffineTransformIdentity };
+
+    // Alpha Blending
+    CGFloat alpha = 1.0f;
+};
+
+struct __CGContext {
+    struct __CGContextImpl {
+        ComPtr<ID2D1RenderTarget> renderTarget{ nullptr };
+
+        // Calculated at creation time, this transform flips CG's drawing commands,
+        // anchored in the bottom left, to D2D's top-left coordinate system.
+        CGAffineTransform cgCompatibilityTransform{ CGAffineTransformIdentity };
+
+        std::stack<__CGContextDrawingState> stateStack{};
+
+        woc::unique_cf<CGMutablePathRef> currentPath{ nullptr };
+
+        // TODO(DH) GH#1070 evaluate these defaults; they should be set by context creators.
+        bool allowsAntialiasing = false;
+        bool allowsFontSmoothing = false;
+        bool allowsFontSubpixelPositioning = false;
+        bool allowsFontSubpixelQuantization = false;
+
+        __CGContextImpl() {
+            // Set up a default/baseline state
+            stateStack.emplace();
+        }
+    };
+
+    CFRuntimeBase _base;
+
+    // Use an inline impl struct so that we don't need to placement new each element.
+    __CGContextImpl _impl;
+
+    inline ComPtr<ID2D1RenderTarget>& RenderTarget() {
+        return _impl.renderTarget;
+    }
+
+    inline std::stack<__CGContextDrawingState>& GStateStack() {
+        return _impl.stateStack;
+    }
+
+    inline __CGContextDrawingState& CurrentGState() {
+        return GStateStack().top();
+    }
+
+    inline void ClearPath() {
+        _impl.currentPath.reset();
+    }
+
+    inline ComPtr<ID2D1Factory> Factory() {
+        ComPtr<ID2D1Factory> factory;
+        _impl.renderTarget->GetFactory(&factory);
+        return factory;
+    }
+};
 
-#define DEBUG_CONTEXT_COUNT
+#pragma region CFRuntimeClass
+static void __CGContextInit(CFTypeRef cf) {
+    CGContextRef context = (CGContextRef)cf;
+    struct __CGContext* mutableContext = const_cast<struct __CGContext*>(context);
+    new (&mutableContext->_impl) __CGContext::__CGContextImpl();
+}
 
-int contextCount = 0;
-static IWLazyClassLookup _LazyUIImage("UIImage"), _LazyUIScreen("UIScreen");
+static void __CGContextDeallocate(CFTypeRef cf) {
+    CGContextRef context = (CGContextRef)cf;
+    context->_impl.~__CGContextImpl();
+}
 
-pthread_mutex_t _cairoLock = PTHREAD_MUTEX_INITIALIZER;
+template <typename T, typename PT = typename std::remove_pointer<T>::type*, typename B = decltype(std::declval<PT>()->_base)>
+static PT _CFRuntimeCreateInstance(CFTypeID typeID, CFAllocatorRef allocator = kCFAllocatorDefault) {
+    return static_cast<PT>(
+        const_cast<void*>(_CFRuntimeCreateInstance(allocator, typeID, sizeof(typename std::remove_pointer<T>::type) - sizeof(B), nullptr)));
```

Comment:
`sizeof(PT)`will always be 4 on our supported platforms. It is definitely not what we want.